transports/quic/src/muxer.rs

```diff
@@ -0,0 +1,360 @@
+                }
+                Ok(None) => break,
+                Err(err @ quinn_proto::ReadError::Reset(_)) => {
+                    return Poll::Ready(Err(io::Error::new(io::ErrorKind::ConnectionReset, err)))
+                }
+                Err(quinn_proto::ReadError::Blocked) => {
+                    pending = true;
+                    break;
+                }
+            }
+        }
+        if chunks.finalize().should_transmit() {
+            if let Some(waker) = muxer.poll_connection_waker.take() {
+                waker.wake();
+            }
+        }
+        if pending && bytes == 0 {
+            let substream_state = muxer
+                .substreams
+                .get_mut(&self.id)
+                .expect("known substream; qed");
+            substream_state.read_waker = Some(cx.waker().clone());
+            Poll::Pending
+        } else {
+            Poll::Ready(Ok(bytes))
+        }
+    }
+}
+
+impl AsyncWrite for Substream {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::io::Result<usize>> {
+        let mut muxer = self.muxer.lock();
+
+        match muxer.connection.send_stream(self.id).write(buf) {
+            Ok(bytes) => Poll::Ready(Ok(bytes)),
+            Err(quinn_proto::WriteError::Blocked) => {
+                let substream = muxer
+                    .substreams
+                    .get_mut(&self.id)
+                    .expect("known substream; qed");
+                substream.write_waker = Some(cx.waker().clone());
+                Poll::Pending
+            }
+            Err(err @ quinn_proto::WriteError::Stopped(_)) => {
+                Poll::Ready(Err(io::Error::new(io::ErrorKind::ConnectionReset, err)))
+            }
+            Err(quinn_proto::WriteError::UnknownStream) => {
+                Poll::Ready(Err(io::ErrorKind::BrokenPipe.into()))
+            }
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+        // quinn doesn't support flushing, calling close will flush all substreams.
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+        let mut muxer = self.muxer.lock();
+        match muxer.connection.finish_substream(self.id) {
+            Ok(()) => {
+                let substream_state = muxer
+                    .substreams
+                    .get_mut(&self.id)
+                    .expect("Substream is not finished.");
+                substream_state.finished_waker = Some(cx.waker().clone());
+                Poll::Pending
+            }
+            Err(err @ quinn_proto::FinishError::Stopped(_)) => {
+                Poll::Ready(Err(io::Error::new(io::ErrorKind::ConnectionReset, err)))
+            }
+            Err(quinn_proto::FinishError::UnknownStream) => Poll::Ready(Ok(())),
+        }
```

Comment:
This is a bit hacky. We are relying on the `UnknownStream` error to figure out whether we have successfully closed the stream once this task gets woken up again because the waker has been called.

We could check for the presence of the stream in our hashmap and if it is not there, return `Poll::Ready(Ok())`?

Reply:
I agree that it is kinda hacky and not ideal. Problem is that a stream could be in our hashmap but still already finished. Finished streams can still be used for reading. Alternative is to explicitly set a flag `is_finished` once the connection returns a `ConnectionEvent::StreamFinished` and then return early here if this flag is set to `true`. Thinking about it again, I think this is easier to understand. Unless you have an alternative idea?

Reply:
Yeah I think I'd prefer to explicitly track this state.

Reply:
Added `is_write_closed` flag in 82b6a9d.